### PR TITLE
Decimal128 Support [databricks]

### PR DESCRIPTION
**Pull request by revans2**
This adds in support for decimal 128 and has been a collaboration with lots of different people.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/v2/Spark30XShims.scala [hunk 1]
```diff
@@ -96,7 +96,7 @@ trait Spark30XShims extends SparkShims {
 
   override def aqeShuffleReaderExec: ExecRule[_ <: SparkPlan] = exec[CustomShuffleReaderExec](
     "A wrapper of shuffle query stage",
-    ExecChecks((TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL_64 + TypeSig.ARRAY +
+    ExecChecks((TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL_128_FULL + TypeSig.ARRAY +
```
Comment (jlowe) [inline sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/v2/Spark30XShims.scala:99]: I think the 312db shim needs this change as well.

**Comment by revans2** [comment]
build

**Review by jlowe — DISMISSED** [review]

**Comment by revans2** [comment]
build

**Review by kuhushukla — DISMISSED** [review]

**Comment by revans2** [comment]
Tests are failing because of some issues with changes to SUM in 3.1.0. I am trying to figure out how to make this work with the older versions of Spark.

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
@jlowe @kuhushukla 

I added in a basic fix for decimal. It is not great because there is a circular dependency between different files now, and I want to break up AggregateFunctions.scala, but that is too large of a change to make here.

**Review by jlowe — DISMISSED** [review]

**Comment by revans2** [comment]
Test failed with hash percentile approximate for decimal.  This is an intermittent failure and is still being investigated #4060

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
It failed on 3.3.0 compile even after nothing changed. It might be a Spark change itself.
```
[2021-11-18T17:32:54.272Z] [ERROR] [Error] /home/jenkins/agent/workspace/jenkins-rapids_premerge-github-3312/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuOrcFileFormat.scala:132: not found: value OrcFileFormat
```

**Comment by revans2** [comment]
Yup.  https://github.com/apache/spark/commit/6f6a066dff7e1ceceb80ab5c3809d5e6fb1474e2 removed the Object OrcFileFormat

**Comment by revans2** [comment]
build

**Review by jlowe — DISMISSED** [review]

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
Same failure as before :(

**Comment by revans2** [comment]
build

**Review by jlowe — APPROVED** [review]

**Comment by revans2** [comment]
This time it was `test_hash_groupby_approx_percentile_long_repeated_keys`. This has a fix that is blocking other things so I am going to just merge it.

**Pull request merged**